warehouse/templates/manage/project/publishing.html

```diff
@@ -146,7 +105,64 @@ <h3>GitHub</h3>
         <input type="submit" value="{% trans %}Add{% endtrans %}" class="button button--primary">
       </div>
     </form>
+{% endmacro %}
+
+{% block main %}
+{% if testPyPI %}
+{% set title = "TestPyPI" %}
+{% else %}
+{% set title = "PyPI" %}
+{% endif %}
 
+<div class="horizontal-section">
+  <div class="site-container">
+    <h1 class="page-title">{{ oidc_title() }}</h1>
+
+    {{ oidc_desc() }}
+
+    <h2>{% trans %}Manage current publishers{% endtrans %}</h2>
+    {% if project.oidc_publishers %}
+    <table class="table table--publisher-list">
+      <caption class="sr-only">
+        {% trans project_name=project.name %}OpenID Connect publishers associated with {{ project_name }}{% endtrans %}
+      </caption>
+      <thead>
+        <tr>
+          <th scope="col">{% trans %}Publisher{% endtrans %}</th>
+          <th scope="col">{% trans %}Details{% endtrans %}</th>
+          <th></th>
+        </tr>
+      </thead>
+      <tbody>
+        {% for publisher in project.oidc_publishers %}
+        {{ oidc_publisher_row(publisher) }}
+        {% endfor %}
+      </tbody>
+    </table>
+    {% else %}
+    <p class="no-bottom-padding">{% trans %}No publishers are currently configured.{% endtrans %}</p>
+    {% endif %}
+
+    <h2 class="no-bottom-padding">{% trans %}Add a new publisher{% endtrans %}</h2>
+
+    {% if request.user.has_two_factor %}
+
+    {% set publishers = [("GitHub", github_form(request, github_publisher_form))] %}
```

Comment:
Agreed, although I'm not sure where best this would live on the Python side. Maybe in the default render for the associated view?

(I can do this in follow-up if the two of you would prefer.)

Reply:
This is a mapping of the publisher name to the result of a Jinja macro that takes the form as input, not the form itself. I also don't really see where this would live on the Python side unfortunately, but since we'll always need some custom per-provider HTML to render the form, I think this is acceptable.